### PR TITLE
8280890: Cannot use '-Djava.system.class.loader' with class loader in signed JAR

### PR DESCRIPTION
**Pull request by seanjmullan**
This fixes a bootstrapping issue if a custom system class loader is set with the `-Djava.system.class.loader` option and the custom class loader is inside a signed JAR.  In order to load the custom class loader, the runtime must verify the signed JAR first, and the algorithm constraint code tries to load a `Locale` provider using a `ServiceLoader` before the class loader is set, and this causes a `ServiceConfigurationError`.

The fix removes a dependency from the security algorithm "denyAfter" constraint parsing code on the `Calendar` API which uses a `ServiceLoader` for gathering default locale information. Instead the `ZonedDateTime` API is now used, which simplifies the code and removes some unnecessary code from `keytool` as well.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280890](https://bugs.openjdk.java.net/browse/JDK-8280890): Cannot use '-Djava.system.class.loader' with class loader in signed JAR

### Reviewers
 * [Weijun Wang](https://openjdk.java.net/census#weijun) (@wangweij - **Reviewer**) ⚠️ Review applies to 914c88214af821af9276128ce423a085270a7345
 * [Hai-May Chao](https://openjdk.java.net/census#hchao) (@haimaychao - Committer) ⚠️ Review applies to 914c88214af821af9276128ce423a085270a7345

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7316/head:pull/7316` \
`$ git checkout pull/7316`

Update a local copy of the PR: \
`$ git checkout pull/7316` \
`$ git pull https://git.openjdk.java.net/jdk pull/7316/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7316`

View PR using the GUI difftool: \
`$ git pr show -t 7316`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7316.diff">https://git.openjdk.java.net/jdk/pull/7316.diff</a>

</details>

### CODE REVIEW

**Review by wangweij — APPROVED** [review]
LGTM.

**Review by haimaychao — APPROVED** [review]

**Comment by haimaychao** [comment]
Looks good to me.

File: test/jdk/java/security/SignedJar/SignedJarWithCustomClassLoader.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+             testClasses);
+
+        // create the jar file
+        Path classes = Paths.get(testClasses);
+        JarUtils.createJarFile(Paths.get("test.jar"), classes,
+            classes.resolve("CustomClassLoader.class"),
+            classes.resolve("Main.class"));
+
+        // create signer's keypair
+        SecurityTools.keytool("-genkeypair -keyalg RSA -keystore ks " +
```
Comment (jaikiran) [inline test/jdk/java/security/SignedJar/SignedJarWithCustomClassLoader.java:65]: Hello Sean,
Looking at the `SecurityTools.keytool` and `SecurityTools.jarsigner` methods, they internally launch a process corresponding to these tools but do not check for the exit code of that process execution. Perhaps the calls to these methods in this test, should add a check to assert that the exit code is `0` by using the returned `OutputAnalyzer`?

Comment (seanjmullan) [reply test/jdk/java/security/SignedJar/SignedJarWithCustomClassLoader.java:65]: Good catch - fixed in latest update.

**Comment by jaikiran** [comment]
Thank you for that change, Sean. Looks fine to me.

**Pull request closed without merging**